File: Capitulo3_VRCardboard - JAVA/app/src/main/java/com/upv/pm_2022/sep_dic/capitulo3_vrcardboard/Distortion.java
package com.upv.pm_2022.sep_dic.capitulo3_vrcardboard;

import java.util.*;

public class Distortion {
    private static final float[] DEFAULT_COEFFICIENTS;
    private float[] mCoefficients;
    
    public Distortion() {
        super();
        this.mCoefficients = Distortion.DEFAULT_COEFFICIENTS.clone();
    }
    
    public Distortion(final Distortion other) {
        super();
        this.setCoefficients(other.mCoefficients);
    }
    
    public static Distortion parseFromProtobuf(final float[] coefficients) {
        final Distortion distortion = new Distortion();
        distortion.setCoefficients(coefficients);
        return distortion;
    }
    
    public float[] toProtobuf() {
        return this.mCoefficients.clone();
    }
    
    public void setCoefficients(final float[] coefficients) {
        this.mCoefficients = ((coefficients != null) ? coefficients.clone() : new float[0]);
    }
    
    public float distortionFactor(final float radius) {
        float result = 1.0f;
        float rFactor = 1.0f;
        final float rSquared = radius * radius;
        for (final float ki : this.mCoefficients) {
            rFactor *= rSquared;
            result += ki * rFactor;
        }
        return result;
    }
    
    public float distort(final float radius) {
        return radius * this.distortionFactor(radius);
    }
    
    public float distortInverse(final float radius) {
        float r0 = radius / 0.9f;
        float r = radius * 0.9f;
        float dr0 = radius - this.distort(r0);
        while (Math.abs(r - r0) > 1.0E-4) {
            final float dr = radius - this.distort(r);
            final float r2 = r - dr * ((r - r0) / (dr - dr0));
            r0 = r;
            r = r2;
            dr0 = dr;
        }
        return r;
    }
    
    @Override
    public boolean equals(final Object other) {
        if (other == null) {
            return false;
        }
        if (other == this) {
            return true;
        }
        if (!(other instanceof Distortion)) {
            return false;
        }
        final Distortion o = (Distortion)other;
        return Arrays.equals(this.mCoefficients, o.mCoefficients);
    }
    
    @Override
    public String toString() {
        final StringBuilder builder = new StringBuilder().append("{\n").append("  coefficients: [");
        for (int i = 0; i < this.mCoefficients.length; ++i) {
            builder.append((this.mCoefficients[i]));
            if (i < this.mCoefficients.length - 1) {
                builder.append(", ");
            }
        }
        builder.append("],\n}");
        return builder.toString();
    }
    
    static {
        DEFAULT_COEFFICIENTS = new float[] { 0.441f, 0.156f };
    }
}

File: VRCardboardKotlin/app/src/main/java/com/vrcardboardkotlin/Distortion.kt
package com.vrcardboardkotlin

import kotlin.math.abs

class Distortion {
    private lateinit var mCoefficients: FloatArray

    constructor() : super() {
        this.mCoefficients = DEFAULT_COEFFICIENTS.clone()
    }

    constructor(other: Distortion?) : super() {
        this.setCoefficients(other!!.mCoefficients)
    }

    fun toProtobuf(): FloatArray {
        return mCoefficients.clone()
    }

    fun setCoefficients(coefficients: FloatArray?) {
        this.mCoefficients = (if ((coefficients != null)) coefficients.clone() else FloatArray(0))
    }

    fun distortionFactor(radius: Float): Float {
        var result = 1.0f
        var rFactor = 1.0f
        val rSquared = radius * radius
        for (ki in this.mCoefficients) {
            rFactor *= rSquared
            result += ki * rFactor
        }
        return result
    }

    fun distort(radius: Float): Float {
        return radius * this.distortionFactor(radius)
    }

    fun distortInverse(radius: Float): Float {
        var r0 = radius / 0.9f
        var r = radius * 0.9f
        var dr0 = radius - this.distort(r0)
        while (abs((r - r0).toDouble()) > 1.0E-4) {
            val dr = radius - this.distort(r)
            val r2 = r - dr * ((r - r0) / (dr - dr0))
            r0 = r
            r = r2
            dr0 = dr
        }
        return r
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) {
            return false
        }
        if (other === this) {
            return true
        }
        if (other !is Distortion) {
            return false
        }
        return mCoefficients.contentEquals(other.mCoefficients)
    }

    override fun toString(): String {
        val builder = StringBuilder().append("{\n").append("  coefficients: [")
        for (i in mCoefficients.indices) {
            builder.append((mCoefficients[i]))
            if (i < mCoefficients.size - 1) {
                builder.append(", ")
            }
        }
        builder.append("],\n}")
        return builder.toString()
    }

    companion object {
        private val DEFAULT_COEFFICIENTS = floatArrayOf(0.441f, 0.156f)
        fun parseFromProtobuf(coefficients: FloatArray?): Distortion {
            val distortion = Distortion()
            distortion.setCoefficients(coefficients)
            return distortion
        }
    }
}
